src/function.ts

```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    if (func && id) {
+      if (isNativeFunctionDecl(func)) {
+        callbackLambdaCode = new CallbackLambdaCode(func.closure);
+        _resource = new aws_lambda.Function(resource, id, {
+          ...props,
+          runtime: aws_lambda.Runtime.NODEJS_14_X,
+          handler: "index.handler",
+          code: callbackLambdaCode,
+        });
+
+        integrations = func.integrations;
+      } else if (isErr(func)) {
+        throw func.error;
+      } else {
+        throw Error(
+          "Expected lambda to be passed a compiled function closure or a aws_lambda.IFunction"
+        );
+      }
+    } else {
+      _resource = resource as aws_lambda.IFunction;
+    }
+    super(_resource);
+
+    // retrieve and bootstrap all found native integrations. Will fail if the integration does not support native integration.
+    const nativeIntegrationsPrewarm = integrations
+      .map((i) => new IntegrationImpl(i))
+      .map((integration) => {
+        const integ = integration.native;
+        integ.bootstrap(this);
+        return integ;
+      })
+      .filter((i) => i.preWarm)
+      .map((i) => i.preWarm);
+    // Start serializing process, add the callback to the promises so we can later ensure completion
+    callbackLambdaCode &&
+      Function.promises.push(
+        callbackLambdaCode.generate(nativeIntegrationsPrewarm)
+      );
+  }
+
+  public static fromFunction<P, O>(
+    func: aws_lambda.IFunction
+  ): ImportedFunction<P, O> {
+    return new ImportedFunction<P, O>(func);
+  }
+}
+
+export class ImportedFunction<P, O> extends FunctionBase<P, O> {
+  constructor(func: aws_lambda.IFunction) {
+    return super(func) as unknown as ImportedFunction<P, O>;
+  }
+}
+
+type ConditionalFunction<P, O> = P extends undefined
+  ? (payload?: P) => O
+  : (payload: P) => O;
+
+/**
+ * A special lambda code wrapper that serializes whatever closure it is given.
+ *
+ * Caveat: Relies on async functions which may not finish when using CDK's app.synth()
+ *
+ * Ensure the generate function's promise is completed using something like Lambda.promises and the `asyncSynth` function.
+ *
+ * Use:
+ * * Initialize the {@link CallbackLambdaCode} `const code = new CallbackLambdaCode()`
+ * * First bind the code to a Function `new aws_lambda.Function(..., { code })`
+ * * Then call generate `const promise = code.generate(integrations)`
+ * *
+ */
+export class CallbackLambdaCode extends aws_lambda.Code {
+  private scope: Construct | undefined = undefined;
+
+  constructor(
+    private func: (preWarmContext: NativePreWarmContext) => AnyFunction
+  ) {
+    super();
+  }
+
+  public bind(scope: Construct): aws_lambda.CodeConfig {
+    this.scope = scope;
+    // Lets give the function something lightweight while we process the closure.
+    return aws_lambda.Code.fromInline(
+      "If you are seeing this in your lambda code, ensure generate is called and then consult the README."
```

Comment:
Can we link to an issue?

Reply:
https://github.com/sam-goodwin/functionless/issues/128